frink: init at 2023-01-05

fehrenbach opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

The Frink programming language and calculator

https://frinklang.org/

I'd quite like this in 22.11, but I can't add the backport label myself.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

fehrenbach committed:
frink: init at 2023-01-31

Co-authored-by: Anderson Torres <torres.anderson.85@protonmail.com>

fehrenbach committed:
maintainers: add stefanfehrenbach

AndersonTorres commented on pkgs/development/tools/frink/default.nix:
```diff
@@ -0,0 +1,35 @@
+{ pkgs }:
+with pkgs;
+let
+  jar = fetchurl {
```
Why?

Just put this as `src`.

fehrenbach replied:
I wanted to also get the units file and functions file for rlwrap to provide tab completion but I haven't gotten around to it, yet. I'm not sure I ever will and I don't otherwise have strong opinions on it, so sure, if you prefer.

AndersonTorres commented on pkgs/development/tools/frink/default.nix:
```diff
@@ -0,0 +1,35 @@
+    url = "https://frinklang.org/frinkjar/frink.jar";
+    sha256 = "sha256-7pYxQnnOjOSBnk0n/Jt78sNisDKTmDqxqmsPakil5ig=";
+  };
+in stdenv.mkDerivation rec {
```
```suggestion
in
stdenv.mkDerivation rec {
```

AndersonTorres commented on pkgs/development/tools/frink/default.nix:
```diff
@@ -0,0 +1,35 @@
+  };
+in stdenv.mkDerivation rec {
+  name = "frink";
+  src = ./.;
```
Do this like ditaa

https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/graphics/ditaa/default.nix

AndersonTorres commented on pkgs/development/tools/frink/default.nix:
```diff
@@ -0,0 +1,35 @@
+with pkgs;
+let
+  jar = fetchurl {
+    url = "https://frinklang.org/frinkjar/frink.jar";
```
It does not have any version string?

fehrenbach replied:
Nope, not that I could find :(
I will contact the author to ask if there are maybe undocumented versioned jars around somewhere.

fehrenbach replied:
Nope, not that I could find :(
I will contact the author to ask if there are maybe undocumented versioned jars around somewhere.

AndersonTorres replied:
How the author announces new releases?

fehrenbach commented on pkgs/development/tools/frink/default.nix:
```diff
@@ -0,0 +1,35 @@
+{ pkgs }:
+with pkgs;
```
I suspect you mean "lazy" rather than "sluggish".

I don't quite understand what you mean by "detail your inputs", though. Can you please elaborate?

AndersonTorres replied:
Instead of using `{pkgs}: with pkgs; ...` you should use ` { jdk, rlwrap }: ...`.

AndersonTorres replied:
P.S.: not my default lang.

AndersonTorres commented:
@fehrenbach, How the author announces new releases?

If the author merely overwrites the file, you can use the Web Archive trick: save the JAR link on https://web.archive.org/save/ and use the generated link as the source.

Indeed I have done this right now:

https://web.archive.org/web/20230128164131/https://frinklang.org/frinkjar/frink.jar

AndersonTorres commented on pkgs/development/tools/frink/default.nix:
```diff
@@ -0,0 +1,33 @@
+stdenv.mkDerivation rec {
+  name = "frink";
+  src = fetchurl {
+    url = "https://frinklang.org/frinkjar/frink.jar";
```
```suggestion
    # Upstream does not provide versioned download links
    url = "https://web.archive.org/web/20230128164131/https://frinklang.org/frinkjar/frink.jar
";
```

AndersonTorres commented on pkgs/development/tools/frink/default.nix:
```diff
@@ -0,0 +1,33 @@
+{ fetchurl, jdk, lib, rlwrap, stdenv }:
+stdenv.mkDerivation rec {
+  name = "frink";
```
```suggestion
  pname = "frink";
  version = "unstable-2023-01-28";
```

This is the date provided by Web Archive, from the `/web/20230128164131/` substring.

AndersonTorres commented:
> I take it pname + version is now preferred over name? 

Yes. This was an ancient convention originated from Python expressions, and it became popular very quickly, being now the standard.

fehrenbach commented:
Thanks, @AndersonTorres! I took the liberty to squash your suggestions into the "frink: init" commit.

I've also changed the `version` to mach the version reported by frink itself (and added a test (though I guess it won't get executed automatically because unfree)). This means the URL doesn't necessarily match the version (depending on when it's generated by archive.org), but that seems better to me than misrepresenting what version of frink the package contains.

AndersonTorres commented on pkgs/development/tools/frink/default.nix:
```diff
@@ -0,0 +1,41 @@
+stdenv.mkDerivation rec {
+  pname = "frink";
+  version = "2023-01-31";
+  src = fetchurl {
```
```suggestion

  src = fetchurl {
```

AndersonTorres commented on pkgs/development/tools/frink/default.nix:
```diff
@@ -0,0 +1,41 @@
+    sha256 = "sha256-xs1FQvFPgeAxscAiwBBP8N8aYe0OlsYbH/vbzzCbYZc=";
+  };
+  dontUnpack = true;
+  buildInputs = [ jdk rlwrap ];
```
```suggestion
  buildInputs = [ jdk rlwrap ];

```

AndersonTorres commented on pkgs/development/tools/frink/default.nix:
```diff
@@ -0,0 +1,41 @@
+    chmod a+x "$out/bin/frink"
+
+    runHook postInstall
+  '';
```
```suggestion
  '';

```

AndersonTorres commented on pkgs/development/tools/frink/default.nix:
```diff
@@ -0,0 +1,41 @@
+    license = licenses.unfree;
+    sourceProvenance = [ sourceTypes.binaryBytecode ];
+    maintainers = [ maintainers.stefanfehrenbach ];
+  };
```
```suggestion
  };

```

AndersonTorres reviewed (changes requested):
Some whitespaces and cosmetic changes

AndersonTorres commented:
> This means the URL doesn't necessarily match the version (depending on when it's generated by archive.org), but that seems better to me than misrepresenting what version of frink the package contains.

OK.

AndersonTorres reviewed (commented):
1. Fix the extra whitespace at lines 7 and 12, as indicated by the bot
2. Squash the commits into one (except the `maintainers` one).
